### ISSUE TITLE
Messaging SDK does not expose service data payloads via Notification extras

### ISSUE THREAD

**Issue by sfuqua** [firebase/firebase-android-sdk#6387]
### [READ] Step 1: Are you in the right place?

Yes, I believe this is an issue with the Android Firebase SDK that can be mitigated in code in this repository.

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: Jellyfish/2023.3.1
  * Firebase Component: Messaging
  * Component version: 23.4.1 (but issue observed in repo's latest)

### [REQUIRED] Step 3: Describe the problem

Notifications automatically posted by the SDK ("display notifications" when the app is backgrounded) do not expose service-provided data payloads to the application when the application queries its pending notifications via Android's `NotificationManager.getActiveNotifications()`.

Looking at `CommonNotificationBuilder.createNotificationInfo` in this repository, we can see [the contentIntent is set](https://github.com/firebase/firebase-android-sdk/blob/2dd9beb5525a5bcadb436ab118cda11945d4a5d5/firebase-messaging/src/main/java/com/google/firebase/messaging/CommonNotificationBuilder.java#L156), but not `extras`.

Proposal - add this line:
```java
builder.setExtras(params.paramsWithReservedKeysRemoved());
```

This would ensure that the posted `Notification` has the same service-provided extras as the content Intent, which means the app can inspect those extras later.

Alternatively, provide some sort of opt-in mechanism in the app's manifest to enable copying specific keys into the `Notification` extras.

#### Steps to reproduce:

1. Have your service post an FCM "display notification" *with* extra data - that is, it has both a `notification` payload and a `data` body - while your app is backgrounded
2. Observe the Firebase SDK automatically intercepts and posts a Notification to the Android notification center
3. In your app, query `NotificationManager.getActiveNotifications()`
4. Attempt to inspect the extra data of one of your pending notifications, and observe that there is no way to do so; Android does not expose the `contentIntent` extras, and the SDK does not populate the `Notification` extras

This breaks scenarios where the service may want to update or cancel an existing notification based on some sort of identifier - there is no way to plumb an identifier into the Notification that the Firebase SDK has posted.

#### Relevant Code:

App code, e.g. in `FirebaseMessagingService.onMessageReceived` or anywhere else we may wish to inspect previously posted Notifications:
```java
        Context appContext = this.getApplicationContext();
        NotificationManager nativeNotificationManager = appContext.getSystemService(NotificationManager.class);
        StatusBarNotification[] deliveredNotifications = nativeNotificationManager.getActiveNotifications();

        // Walk delivered notifications to find one matching our criteria to update or cancel
        for (StatusBarNotification notification : deliveredNotifications) {
            Notification notificationPayload = notification.getNotification();
            // Here notificationPayload.extras is accessible, but has no useful data because FCM SDK did not include it when creating the Notification
            // notificationPayload.contentIntent exists, but we cannot inspect the extras so it is not useful to us
```

See related issue with confused developer:

* https://github.com/firebase/quickstart-android/issues/1004

**Comment by sfuqua** [comment]
The tl;dr of my scenario is this:

1. My service delivers display notifications with some sort of identifier
2. I later want to update one of these notifications - Android lets me do this by inspecting the `StatusBarNotifications`, finding their tag + ID, and submitting a new notification with those (or cancelling)

The second point is straightforward *if* I can affirmatively identify which `StatusBarNotification` to mutate. Unfortunately the display notifications posted by the SDK do not include my service-side identifiers in any way that I can find them.

`NotificationBuilder.setExtras` or `addExtras` is supposed to be the entrypoint to the Android API to do this (so that you can find your extras later), but the SDK does not call this code and there's no way to enforce the plumbing I need.

**Comment by lehcar09** [comment]
Hi @sfuqua, thank you for reaching out and reporting the issue. For context, App's behavior when receiving messages that include both notification and data payloads depends on whether the app is in the background or the foreground—essentially, whether or not it is active at the time of receipt.

* When in the background, apps receive the notification payload in the notification tray, and only handle the data payload when the user taps on the notification.
* When in the foreground, your app receives a message object with both payloads available.

The `data` accompanying the notification message is available in the intent extras. 
https://github.com/firebase/firebase-android-sdk/blob/2f157570a01463d871191975d67afb784efde7fc/firebase-messaging/src/main/java/com/google/firebase/messaging/CommonNotificationBuilder.java#L385

You can check our [quickstart app](https://github.com/firebase/quickstart-android/blob/4d1e253dd1dd8052310e889fc1785084b1997aca/messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/MainActivity.kt#L65-L70) for sample code snippets on how to get the intent extras.

It looks to me that this is working as intended. That said, I'll be closing this issue. I would suggest using [FCM collapsible messages](https://firebase.google.com/docs/cloud-messaging/concept-options#collapsible_and_non-collapsible_messages) to update the message you’ve sent. 

> Collapsible Messages: When there is a newer message that renders an older, related message irrelevant to the client app, FCM replaces the older message. For example: messages used to initiate a data sync from the server, or outdated notification messages.

Let me know if there's any misunderstanding so we can re-open the issue for further investigation. Thanks!

**Comment by sfuqua** [comment]
Hi @lehcar09, thanks for your time and detailed reply.

I agree regarding the behavior dichotomy. Your first bullet ("only handle the data payload when the user taps on the notification") is indeed how the Firebase Android SDK behaves today, and that the data payload is available in the `Intent` extras, and my app is handling that case well today.

I also appreciate your suggestion to investigate collapsible messages. That is something I did before opening this issue. Please believe me when I say I am very well acquainted with FCM documentation as well as all of the Java code in the `messaging` component of the SDK as I have been working very intimately with it for a few years.

Unfortunately collapsible messages are not a workaround here for a couple reasons:

1. They do not allow *canceling* messages. The Android platform allows canceling Notifications via `NotificationManager.cancel`, but without useful identifiers exposed by the Firebase client SDK, this is not a supported scenario. An app's only choice here is to inspect the visual attributes of the posted `Notifications` (e.g., the title/body strings), which is not reliable especially when dealing with localized content.
    * An example of this being useful is in updates like a chat or email app, where you might want to cancel a delivered notification when the user has "marked as read" the content on another device.
3. Collapsible messages do not allow updating *delivered* messages, as they are a Firebase service-side concept. Imagine the scenario where the user receives notifications overnight, or while their phone is on but unused (e.g., during a movie) and is receiving push payloads - it may be useful to update those payloads before the user sees them but after they are delivered. Again, Android as a platform supports this, the Firebase Android SDK (this repo) does not.

This issue is another example of the Firebase SDK's "convenience" behavior of handling display notifications actually being (forgive the words here, and with all due respect) an "anti-feature". The SDK constructs a `NotificationBuilder` and delivers to the Android platform without calling back into the host at all, which means:

1. The app cannot configure the `extras` on the posted `Notification`
2. The app cannot override styling choices except the limited amount that the SDK allows configuring through the AndroidManifest
4. The app cannot take advantage of features like `Actions` for inline accelerators on notifications

See also #2639, which is related.

I understand that Google seems to have strong opinions about not allowing the app to *override* the behavior here (again, see linked issue and related issues and related PR that I have previously introduced to this repo), however, I am hopeful that we can discuss tactical improvements like configuring `extras`.

Ideally, the Firebase SDK would call back into the app with the `NotificationBuilder` before posting the Notification, for example via a `virtual` function on the `FirebaseMessagingService`.

In the absence of this level of configurability, Android devs are sort of stuck with a bad choice - to work around these design decisions in the SDK the "best" option always seems to be to forego display notifications altogether and *only* use data-only payloads if your app needs *any* sort of dynamic configurability of Notifications, which would be unfortunate.
